### DISCUSSION TITLE
Support for schema

### DISCUSSION THREAD

**Discussion by nlang** [permitio/cedar-agent#7] [Q&A] [answered]
The Cedar Specification defines support for Schema definitions to validate against when defining policies (see [Cedar Documentation](https://docs.cedarpolicy.com/schema.html))

I couldn't find endpoints to define Schema or anything about it in the current API of cedar-agent. Is there a way schema can be used right now or, if not, are there any plans to support this in the future?

Thanks,
Nicolai

**Comment by asafc** [comment, marked as answer]
Hi @nlang, i don't think we implemented validation schemas as of yet.
@omer9564 please correct me if i'm wrong.

We definitely plan to support this in the near future. We are currently doing some extensive work on OPAL and after that we are planning to expand Cedar-Agent with more features. If you wish to contribute this feature sooner, our main author @omer9564 is happy to guide you.

**Reply by omer9564** [reply]
You are right, we implemented the "mission critical" parts to support authorization queries.

**Reply by tisauro** [reply]
Hi @omer9564 is there a specification of the work you planning to do to introduce schema validation?

**Reply by omer9564** [reply]
Not yet, I'll check with @asafc and @orweis next week about prioritization of this feature and get back to you 🙂

**Reply by obsd** [reply]
Hi @tisauro and @nlang, we checked our roadmap, and it will take us time to tackle this issue.
If you consider to contribute this feature, we and the entire community will appreciate it 👍 
Let us know if this is the case and we will be happy to help you with guidance and code review.

**Reply by tisauro** [reply]
Hi @obsd can you please provide some sort of specification to help me out to understand the effort required?

**Reply by omer9564** [reply]
Sure, what you need to do is:
- add another type of "store". Currently there is MemoryPolicyStore and MemoryDataStore ( for entities ) - add another one for the schema.
- add an api routes to get, update and delete the schema ( when updating you need to verify that the policy and the entities match the schema ).
- use the schema when updating and creating policy and data.
